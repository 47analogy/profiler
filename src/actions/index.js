export function waitingForProfileFromAddon() {
  return {
    type: 'WAITING_FOR_PROFILE_FROM_ADDON'
  };
}

export function receiveProfileFromAddon(profile) {
  return {
    type: 'RECEIVE_PROFILE_FROM_ADDON',
    profile: profile
  };
}

export function requestingSymbolTable(requestedLib) {
  return {
    type: 'REQUESTING_SYMBOL_TABLE',
    requestedLib
  };
}

export function receivedSymbolTableReply(requestedLib) {
  return {
    type: 'RECEIVED_SYMBOL_TABLE_REPLY',
    requestedLib
  };
}

export function startSymbolicating() {
  return {
    type: 'START_SYMBOLICATING'
  };
}

export function profileSymbolicationStep(profile) {
  return {
    type: 'PROFILE_SYMBOLICATION_STEP',
    profile
  };
}

export function doneSymbolicating() {
  return { type: 'DONE_SYMBOLICATING' };
}

export function mergeFunctions(threadIndex, oldFuncToNewFuncMap) {
  return {
    type: 'MERGE_FUNCTIONS',
    threadIndex, oldFuncToNewFuncMap
  };
}

export function assignFunctionNames(threadIndex, funcIndices, funcNames) {
  return {
    type: 'ASSIGN_FUNCTION_NAMES',
    threadIndex, funcIndices, funcNames
  };
}

export function changeSelectedFuncStack(selectedFuncStack) {
  return {
    type: 'CHANGE_SELECTED_FUNC_STACK',
    selectedFuncStack
  };
}
