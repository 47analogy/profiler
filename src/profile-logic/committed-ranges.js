/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
// @flow

import type { StartEndRange } from 'firefox-profiler/types';

/**
 * Users can make preview range selections on the profile, and then can commit these
 * to drill down into a profile. This file contains functions for working with these
 * committed ranges.
 */

/**
 * Parse URL encoded committed ranges with the form:
 * "<start><unit><duration>~<start><unit><duration>", where `start` and
 * `duration` are both integer numbers expressed with the specified unit.
 * `start` can be negative.
 */
export function parseCommittedRanges(
  stringValue: string = ''
): StartEndRange[] {
  if (!stringValue) {
    return [];
  }
  return (
    stringValue
      .split('~')
      .map(s => {
        // Strings look like: 12345m25, which means the start is at 12'345ms, and
        // the duration is 25ms. All values are integer.
        // For microseconds: 12345678u25: the start is at 12'345'678µs, and the
        // duration is 25µs.
        // For nanoseconds: 12345678901n25: the start is at 12'345'678'901ns, and the
        // duration is 25ns.
        const m = s.match(/^(-?[0-9]+)([mun])([0-9]+)$/);
        if (!m) {
          // Flow doesn't like that we return null, but that's OK
          // because we filter it out in the last operation.
          // $FlowExpectError
          return null;
        }

        // Let's convert values to milliseconds.
        const start = Number(m[1]);
        const durationUnit = m[2];
        const duration = Number(m[3]);

        let startInMs;
        let endInMs;
        switch (durationUnit) {
          case 'm':
            // values are in milliseconds.
            startInMs = start;
            endInMs = start + duration;
            break;
          case 'u':
            // values are in microseconds.
            startInMs = start / 1000;
            endInMs = (start + duration) / 1000;
            break;
          case 'n':
            // values are in nanoseconds.
            startInMs = start / 1000000;
            endInMs = (start + duration) / 1000000;
            break;
          default:
            throw new Error(
              `We couldn't recognize the unit ${durationUnit}, which can't happen because the regexp wouldn't match.`
            );
        }

        if (startInMs === endInMs) {
          // Duration could be 0 at first, or float operations rounding errors could
          // also produce this: make sure that we have a non-empty range.
          endInMs += 0.0001;
        }

        return { start: startInMs, end: endInMs };
      })
      // Filter out possible null values coming from bad inputs.
      .filter(value => value)
  );
}

// This function returns a string representation of the pair { start, end } in
// the form of the start + a duration: "<start><unit><duration>". We may lose
// some precision, which is OK. We just need to take care that the resulting
// range is close in length _and_ includes the requested range.
// Note that start and end inputs are in milliseconds.
export function stringifyStartEnd({ start, end }: StartEndRange): string {
  // Let's work in integer nanoseconds for calculations, and with string
  // manipulations, so that we avoid rounding errors due to floats.
  const startInNs = Math.floor(start * 1000000);
  const endInNs = Math.ceil(end * 1000000);

  const durationInNs = endInNs - startInNs;
  const strDurationInNs = String(durationInNs);

  let result;
  if (durationInNs > 9000000) {
    // If the initial duration is more than 9ms, we'll output milliseconds
    const startInMs = String(startInNs).slice(0, -6);
    let durationInMs = strDurationInNs.slice(0, -6);
    if (!strDurationInNs.endsWith('000000')) {
      // We round up the duration, this is like Math.ceil on integer parts.
      durationInMs = Number(durationInMs) + 1;
    }
    result = `${startInMs}m${durationInMs}`;
  } else if (durationInNs > 9000) {
    // If the initial duration is more than 9µs, we'll output microseconds
    const startInUs = String(startInNs).slice(0, -3);
    let durationInUs = strDurationInNs.slice(0, -3);
    if (!strDurationInNs.endsWith('000')) {
      // We round up the duration, this is like Math.ceil on integer parts.
      durationInUs = Number(durationInUs) + 1;
    }
    result = `${startInUs}u${durationInUs}`;
  } else if (durationInNs === 0) {
    result = `${startInNs}n1`;
  } else {
    result = `${startInNs}n${strDurationInNs}`;
  }

  return result;
}

/**
 * Stringify committed ranges into the following form: "start-end~start-end".
 */
export function stringifyCommittedRanges(
  arrayValue: StartEndRange[] = []
): string {
  return arrayValue.map(stringifyStartEnd).join('~');
}

export function getFormattedTimeLength(length: number): string {
  if (length >= 10000) {
    return `${(length / 1000).toFixed(0)} sec`;
  }
  if (length >= 1000) {
    return `${(length / 1000).toFixed(1)} sec`;
  }
  return `${length.toFixed(0)} ms`;
}

export function getCommittedRangeLabels(
  committedRanges: StartEndRange[]
): string[] {
  const labels = committedRanges.map(range =>
    getFormattedTimeLength(range.end - range.start)
  );
  return labels;
}
