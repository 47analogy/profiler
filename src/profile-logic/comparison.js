/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
// @flow

/*
 * This file contains all functions that are needed to achieve profiles
 * comparison: how to merge profiles, how to diff them, etc.
 */

import { stripIndent } from 'common-tags';
import type { IndexIntoCategoryList, CategoryList } from '../types/profile';

type TranslationMapForCategories = Map<
  IndexIntoCategoryList,
  IndexIntoCategoryList
>;

/**
 * Merges several categories lists into one, resolving duplicates if necessary.
 * It returns a translation map that can be used in `adjustCategories` later.
 */
export function mergeCategories(
  categoriesPerThread: CategoryList[]
): {|
  categories: CategoryList,
  translationMaps: TranslationMapForCategories[],
|} {
  const newCategories = [];
  const translationMaps = [];
  const insertedCategories: Map<string, IndexIntoCategoryList> = new Map();

  categoriesPerThread.forEach(categories => {
    const translationMap = new Map();
    translationMaps.push(translationMap);

    categories.forEach((category, i) => {
      const { name } = category;
      const insertedCategoryIndex = insertedCategories.get(name);
      if (insertedCategoryIndex !== undefined) {
        translationMap.set(i, insertedCategoryIndex);
        return;
      }

      translationMap.set(i, newCategories.length);
      insertedCategories.set(name, newCategories.length);
      newCategories.push(category);
    });
  });

  return { categories: newCategories, translationMaps };
}

/**
 * Adjusts the category indices in a category list using a translation map.
 */
export function adjustCategories(
  categories: $ReadOnlyArray<IndexIntoCategoryList | null>,
  translationMap: TranslationMapForCategories
): Array<IndexIntoCategoryList | null> {
  return categories.map(category => {
    if (category === null) {
      return null;
    }
    const result = translationMap.get(category);
    if (result === undefined) {
      throw new Error(
        stripIndent`
          Category with index ${category} hasn't been found in the translation map.
          This shouldn't happen and indicates a bug in perf-html's code.
        `
      );
    }
    return result;
  });
}
