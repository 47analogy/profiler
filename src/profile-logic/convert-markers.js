/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

export function upgradeGCMinorMarker(marker: Object) {
  if ('nursery' in marker) {
    if ('status' in marker.nursery) {
      if (marker.nursery.status === 'no collection') {
        marker.nursery.status = 'nursery empty';
      }
    } else {
      /*
       * This is the old format for GCMinor, rename some
       * properties to the more sensible names in the newer
       * format and set the status.
       *
       * Note that we don't delete certain properties such as
       * promotion_rate, leave them so that anyone opening the
       * raw json data can still see them in converted profiles.
       */
      marker.nursery.status = 'complete';

      marker.nursery.bytes_used = marker.nursery.nursery_bytes;
      delete marker.nursery.nursery_bytes;

      // cur_capacity cannot be filled in.
      marker.nursery.new_capacity = marker.nursery.new_nursery_bytes;
      delete marker.nursery.new_nursery_bytes;

      marker.nursery.phase_times = marker.nursery.timings;
      delete marker.nursery.timings;
    }
  }
}

/*
 * Fix the units for GCMajor and GCSlice phase times.
 */
export function convertPhaseTimes(old_phases: Object): Object {
  const phases = {};
  for (const phase in old_phases) {
    phases[phase] = old_phases[phase] * 1000;
  }
  return phases;
}

/*
 * Upgrade a GCMajor marker in the Gecko profile format.
 */
export function upgradeGCMajorMarker_Gecko8To9(marker: Object) {
  if ('timings' in marker) {
    if (!('status' in marker.timings)) {
      /*
       * This is the old version of the GCMajor marker.
       */

      const timings = marker.timings;

      timings.status = 'completed';

      /*
       * The old version had a bug where the slices field could be included
       * twice with different meanings.  So we attempt to read it as either
       * the number of slices or a list of slices.
       */
      if (Array.isArray(timings.sices)) {
        timings.slices_list = timings.slices;
        timings.slices = timings.slices.length;
      }

      timings.allocated_bytes = timings.allocated * 1024 * 1024;
    }
  }
}

export function upgradeGCMajorMarker_Processed8to9(marker: Object) {
  upgradeGCMajorMarker_Gecko8To9(marker);
  if ('timings' in marker) {
    const timings = marker.timings;
    if (timings.status === 'completed') {
      /*
       * Processed profiles - convert these fields to have more sensible
       * units and sometimes names.
       */
      timings.phase_times = convertPhaseTimes(timings.totals);
      delete timings.totals;

      timings.mmu_20ms /= 100;
      timings.mmu_50ms /= 100;
    }
  }
}
