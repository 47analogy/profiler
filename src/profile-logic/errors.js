/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import type { Library } from './symbol-store';

// Used during the symbolication process to express that we couldn't find
// symbols for a specific library
export class SymbolsNotFoundError extends Error {
  library: Library;
  error: ?Error;

  constructor(message: string, library: Library, error?: Error) {
    super(message);
    this.name = 'SymbolsNotFoundError';
    this.library = library;
    this.error = error;
  }
}
