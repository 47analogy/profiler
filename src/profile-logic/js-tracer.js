/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
// @flow

import type { JsTracerTable } from '../types/profile';
import type { JsTracerTiming } from '../types/profile-derived';
import type { Microseconds } from '../types/units';
import type { UniqueStringArray } from '../utils/unique-string-array';
/**
 * This function is very similar in implementation as getStackTimingByDepth.
 * It creates a list of JsTracerTiming entries that represent the underlying
 * tree structure of the tracing data. This is then used by the JS Tracer panel
 * to render the chart to the screen.
 *
 * The data looks visually something like this:
 *
 * [A------------------------]
 *    [B------][F----]  [H--]
 *       [C---] [G]
 *       [D][E]
 *
 * Where a single row, like B F H, would be one JsTracerTiming.
 */
export function getJsTracerTiming(
  jsTracer: JsTracerTable,
  stringTable: UniqueStringArray
): JsTracerTiming[] {
  const jsTracerTiming: JsTracerTiming[] = [];

  // Go through all of the events.
  for (
    let tracerEventIndex = 0;
    tracerEventIndex < jsTracer.length;
    tracerEventIndex++
  ) {
    const stringIndex = jsTracer.events[tracerEventIndex];
    const displayName = stringTable.getString(stringIndex);

    // Place the event in the closest row that is empty.
    for (let i = 0; i <= jsTracerTiming.length; i++) {
      // Get or create a row.
      let timingRow = jsTracerTiming[i];
      if (!timingRow) {
        timingRow = {
          start: [],
          end: [],
          index: [],
          label: [],
          name: 'Tracing Information',
          length: 0,
        };
        jsTracerTiming.push(timingRow);
      }

      // The timing is converted here from Microseconds to Milliseconds.
      const division = 1000;
      const start = jsTracer.timestamps[tracerEventIndex] / division;
      const durationRaw = jsTracer.durations[tracerEventIndex];
      const duration = durationRaw === null ? 0 : durationRaw / division;

      // Since the events are sorted, look at the last added event in this row. If
      // the new event fits, go ahead and insert it.
      const otherEnd = timingRow.end[timingRow.length - 1];
      if (otherEnd === undefined || otherEnd <= start) {
        timingRow.start.push(start);
        timingRow.end.push(start + duration);
        timingRow.label.push(displayName);
        timingRow.index.push(tracerEventIndex);
        timingRow.length++;
        break;
      }
    }
  }

  return jsTracerTiming;
}

/**
 * Determine the self time for JS Tracer events. This generates a row of timing
 * information for each type event. This function walks the stack structure of JS Tracer
 * events, and determines what's actually executing at a given time.
 *
 * The data visually would look something like this:
 *
 * A: [----]          [-----]
 * B:       [-----]
 * C:              [-]       [--]
 * D:                            [----]
 *
 * Where "A" would be the name of the event, and the boxes in that row would be the self
 * time. Each row is stored as a JsTracerTiming.
 */
export function getJsTracerLeafTiming(
  jsTracer: JsTracerTable,
  stringTable: UniqueStringArray
): JsTracerTiming[] {
  // Each event type will have it's own timing information, later collapse these into
  // a single array.
  const jsTracerTimingMap: Map<string, JsTracerTiming> = new Map();
  const isUrl = stringTable._array.map(string => /:\/\//.test(string));

  function reportSelfTime(
    tracerEventIndex: number,
    start: Microseconds,
    end: Microseconds
  ): void {
    if (start === end) {
      // If this self time is 0, do not report it.
      return;
    }
    const stringIndex = jsTracer.events[tracerEventIndex];
    const displayName = stringTable.getString(stringIndex);
    const rowName = isUrl[stringIndex] ? 'Script' : displayName;
    let timingRow = jsTracerTimingMap.get(rowName);
    if (timingRow === undefined) {
      timingRow = {
        start: [],
        end: [],
        index: [],
        label: [],
        name: rowName,
        length: 0,
      };
      jsTracerTimingMap.set(rowName, timingRow);
    }
    // Convert the timing to milliseconds.
    const division = 1000;
    if (end < start) {
      throw new Error('end is less than the start');
    }
    timingRow.start.push(start / division);
    timingRow.end.push(end / division);
    timingRow.label.push(displayName);
    timingRow.index.push(tracerEventIndex);
    timingRow.length++;

    if (timingRow.length > 1) {
      const currStart = timingRow.start[timingRow.length - 1];
      const currEnd = timingRow.end[timingRow.length - 1];
      const prevEnd = timingRow.end[timingRow.length - 2];
      if (currEnd < currStart) {
        throw new Error(
          `currEnd < currStart "${displayName} - ${currEnd} < ${currStart}"`
        );
      }
      if (currStart < prevEnd) {
        throw new Error(
          `currStart < prevEnd "${displayName} - ${currStart} < ${prevEnd}"`
        );
      }
      if (currEnd < prevEnd) {
        throw new Error(
          `currEnd < prevEnd "${displayName} - ${currEnd} < ${prevEnd}"`
        );
      }
    }
  }

  // Determine the self time of the various events. These values are all the stack
  // of "prefix" events. The event details will be pushed on to these stacks until
  // they are clearly self time.
  const prefixesStarts = [];
  const prefixesEnds = [];
  const prefixesEventIndexes = [];

  // If there is no prefix, set it to -1. This simplifies some of the real-time
  // type checks that would be required by Flow for this mutable variable.
  let prefixesTip = -1;

  // Go through all of the events. Each `if` branch is documented with a small diagram
  // that includes a little bit of ascii art to help explain the step.
  //
  // Legend:
  // xxxxxxxx - Already reported information, not part of the prefix stack.
  // [======] - Some event on the stack that has not been reported.
  // [prefix] - The prefix event to consider, the top of the prefix stack. This
  //            could also be only part of an event, that has been split into multiple
  //            pieces.
  // [current] - The current event to add.

  for (
    let currentEventIndex = 0;
    currentEventIndex < jsTracer.length;
    currentEventIndex++
  ) {
    const currentStart = jsTracer.timestamps[currentEventIndex];
    const durationRaw = jsTracer.durations[currentEventIndex];
    const duration = durationRaw === null ? 0 : durationRaw;
    const currentEnd = currentStart + duration;

    if (prefixesTip === -1) {
      // Nothing has been added yet, add this "current" event to to the stack of prefixes.
      prefixesTip = 0;
      prefixesStarts[prefixesTip] = currentStart;
      prefixesEnds[prefixesTip] = currentEnd;
      prefixesEventIndexes[prefixesTip] = currentEventIndex;
      continue;
    }

    while (prefixesTip >= 0) {
      const prefixStart = prefixesStarts[prefixesTip];
      const prefixEnd = prefixesEnds[prefixesTip];
      const prefixEventIndex = prefixesEventIndexes[prefixesTip];

      if (prefixEnd <= currentStart) {
        // In this case, the "current" event has passed the other "prefix" event.
        //
        //    xxxxxxxxxxxxxxxx[================]
        //    xxxxxxxx[======]     [current]
        //    [prefix]
        //
        // This next step would match too:
        //
        //    xxxxxxxxxxxxxxxx[================]
        //    xxxxxxxx[prefix]     [current]
        //    xxxxxxxx

        // Commit the previous "prefix" event, then continue on with this loop.
        reportSelfTime(prefixEventIndex, prefixStart, prefixEnd);

        // Move the tip towards the prefix.
        if (process.env.NODE_ENV === 'development') {
          // To aid in debugging locally, reset the values to -1 when they are no
          // longer valid. In production don't do this extra work.
          prefixesEventIndexes[prefixesTip] = -1;
          prefixesStarts[prefixesTip] = -1;
          prefixesEnds[prefixesTip] = -1;
        }
        prefixesTip--;

        if (prefixesTip === -1) {
          // This next step would match too:
          //
          //   [prefix]     [current]

          // There are no more "prefix" events to commit. Add the "current" event on, and
          // break out of this loop.
          prefixesTip = 0;
          prefixesEventIndexes[prefixesTip] = currentEventIndex;
          prefixesStarts[prefixesTip] = currentStart;
          prefixesEnds[prefixesTip] = currentEnd;
          break;
        }

        // This is the only branch that has a `continue`, and not a `break`.
        continue;
      }

      if (prefixEnd > currentEnd) {
        // The current event occludes the prefix event.
        //
        //   [prefix=================]
        //           [current]
        //
        // Split the reported self time of the prefix.
        //
        //   [prefix]xxxxxxxxx[prefix]
        //           [current]
        //
        //                    ^leave the second prefix piece on the "prefixes" stack
        //   ^report the first prefix piece
        //
        //   After reporting we are left with:
        //
        //   xxxxxxxxxxxxxxxxx[======]
        //           [current]

        // Report the first part of the prefix's self time.
        reportSelfTime(prefixEventIndex, prefixStart, currentStart);

        // Shorten the prefix's start time.
        prefixesStarts[prefixesTip] = currentEnd;

        // Now add on the "current" event to the stack of prefixes.
        prefixesTip++;
        prefixesStarts[prefixesTip] = currentStart;
        prefixesEnds[prefixesTip] = currentEnd;
        prefixesEventIndexes[prefixesTip] = currentEventIndex;
        break;
      }

      if (prefixEnd === currentEnd) {
        if (prefixStart !== currentStart) {
          // The current event splits the event above it, so split the reported self
          // time of the prefix.
          //
          //   [prefix]xxxxxxxxx
          //           [current]

          // Report the prefix's self time.
          reportSelfTime(prefixEventIndex, prefixStart, currentStart);

          // Update both the index and the start time.
          prefixesStarts[prefixesTip] = currentStart;
          prefixesEventIndexes[prefixesTip] = currentEventIndex;
        } else {
          // The prefix and current events completely match, so don't report any
          // self time from the prefix. Replace the prefix's index.
          prefixesEventIndexes[prefixesTip] = currentEventIndex;
        }
        break;
      }

      // The data appears to be malformed, report a nice error to the console.
      const prefixName = stringTable.getString(
        jsTracer.events[prefixEventIndex]
      );
      const currentName = stringTable.getString(
        jsTracer.events[currentEventIndex]
      );
      console.error('Current JS Tracer information:', {
        jsTracer,
        stringTable,
        prefixEventIndex,
        currentEventIndex,
      });
      console.error(
        `Prefix (parent) times for "${prefixName}": ${prefixStart}ms to ${prefixEnd}ms`
      );
      console.error(
        `Current (child) times for "${currentName}": ${currentStart}ms to ${currentEnd}ms`
      );
      throw new Error(
        'The JS Tracer information was malformed. Some event lasted longer than its parent event.'
      );
    }
  }

  // Drain off the remaining "prefixes" from the stack, and report the self time.
  for (; prefixesTip >= 0; prefixesTip--) {
    reportSelfTime(
      prefixesEventIndexes[prefixesTip],
      prefixesStarts[prefixesTip],
      prefixesEnds[prefixesTip]
    );
  }

  // Return the list of events, sorted alphabetically.
  return [...jsTracerTimingMap.values()].sort(
    (a, b) => (a.name > b.name ? 1 : -1)
  );
}
