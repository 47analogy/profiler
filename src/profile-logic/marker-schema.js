/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
// @flow
import {
  formatNumber,
  formatPercent,
  formatBytes,
  formatSeconds,
  formatMilliseconds,
  formatTimestamp,
  formatMicroseconds,
  formatNanoseconds,
} from '../utils/format-numbers';
import type { MarkerFormatType } from 'firefox-profiler/types';

/**
 * TODO - These will eventually be stored in the profile, but for now
 * define them here.
 */
export const markerSchema = [
  {
    name: 'Bailout',
    display: ['marker-chart', 'marker-table'],
    data: [
      { key: 'bailoutType', label: 'Type', format: 'string' },
      { key: 'where', label: 'Where', format: 'string' },
      { key: 'script', label: 'Script', format: 'string' },
      { key: 'functionLine', label: 'Function Line', format: 'integer' },
      { key: 'bailoutLine', label: 'Bailout Line', format: 'integer' },
    ],
  },
  {
    name: 'GCMajor',
    display: ['marker-chart', 'marker-table', 'timeline-memory'],
    data: [
      // Use custom handling
    ],
  },
  {
    name: 'GCMinor',
    display: ['marker-chart', 'marker-table', 'timeline-memory'],
    data: [
      // Use custom handling
    ],
  },
  {
    name: 'GCSlice',
    display: ['marker-chart', 'marker-table', 'timeline-memory'],
    data: [
      // Use custom handling
    ],
  },
  {
    name: 'CC',
    label: 'Cycle Collect',
    display: ['marker-chart', 'marker-table', 'timeline-memory'],
    data: [],
  },
  {
    name: 'FileIO',
    display: [],
    data: [
      {
        key: 'operation',
        label: 'Operation',
        format: 'string',
        searchable: true,
      },
      {
        key: 'source',
        label: 'Source',
        format: 'string',
        searchable: true,
      },
      {
        key: 'filename',
        label: 'Filename',
        format: 'file-path',
        searchable: true,
      },
    ],
  },
  {
    name: 'MediaSample',
    display: ['marker-chart', 'marker-table'],
    data: [
      {
        key: 'sampleStartTimeUs',
        label: 'Sample start time',
        format: 'microseconds',
      },
      {
        key: 'sampleEndTimeUs',
        label: 'Sample end time',
        format: 'microseconds',
      },
    ],
  },
  {
    name: 'Styles',
    display: ['marker-chart', 'marker-table'],
    data: [
      {
        key: 'elementsTraversed',
        label: 'Elements traversed',
        format: 'integer',
      },
      { key: 'elementsStyled', label: 'Elements styled', format: 'integer' },
      { key: 'elementsMatched', label: 'Elements matched', format: 'integer' },
      { key: 'stylesShared', label: 'Styles shared', format: 'integer' },
      { key: 'stylesReused', label: 'Styles reused', format: 'integer' },
    ],
  },
  {
    name: 'PreferenceRead',
    display: ['marker-chart', 'marker-table'],
    data: [
      { key: 'prefName', label: 'Name', format: 'string' },
      { key: 'prefKind', label: 'Kind', format: 'string' },
      { key: 'prefType', label: 'Type', format: 'string' },
      { key: 'prefValue', label: 'Value', format: 'string' },
    ],
  },
  {
    name: 'Invalidation',
    display: ['marker-chart', 'marker-table'],
    data: [
      { key: 'url', label: 'URL', format: 'url' },
      { key: 'line', label: 'Line', format: 'integer' },
    ],
  },
  {
    name: 'UserTiming',
    display: ['marker-chart', 'marker-table'],
    data: [{ key: 'name', label: 'Name', format: 'string' }],
  },
  {
    name: 'Text',
    display: ['marker-chart', 'marker-table'],
    data: [{ key: 'name', label: 'Details', format: 'string' }],
  },
  {
    name: 'Log',
    display: ['marker-table'],
    data: [
      { key: 'module', label: 'Module', format: 'string' },
      { key: 'name', label: 'Name', format: 'string' },
    ],
  },
  {
    name: 'tracing',
    display: ['marker-chart', 'marker-table'],
    data: [
      // This is really the "type" of the marker.
      { key: 'category', label: 'Category', format: 'string' },
    ],
  },
  {
    name: 'IPC',
    display: ['marker-chart', 'marker-table'],
    data: [
      { key: 'messageType', label: 'Type', format: 'string' },
      { key: 'sync', label: 'Sync', format: 'string' },
    ],
  },
];

/**
 * This function takes the intended marker schema for a marker field, and applies
 * the appropriate formatting function.
 */
export function formatFromMarkerSchema(
  markerType: string,
  format: MarkerFormatType,
  value: any
): string | null {
  switch (format) {
    case 'url':
    case 'file-path':
    case 'string':
      return value;
    case 'duration':
    case 'time':
      return formatTimestamp(value);
    case 'seconds':
      return formatSeconds(value);
    case 'milliseconds':
      return formatMilliseconds(value);
    case 'microseconds':
      return formatMicroseconds(value);
    case 'nanoseconds':
      return formatNanoseconds(value);
    case 'bytes':
      return formatBytes(value);
    case 'integer':
      return formatNumber(value, 0, 0);
    case 'decimal':
      return formatNumber(value);
    case 'percentage':
      return formatPercent(value);
    default:
      console.error(
        `A marker schema of type "${markerType}" had an unknown format "${(format: empty)}"`
      );
      return value;
  }
}
