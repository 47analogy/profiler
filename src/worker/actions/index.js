import { summarizeProfile } from '../../common/summarize-profile';

export function processProfileSummary() {
  return function (dispatch, getState) {
    dispatch({
      toContent: true,
      type: 'PROFILE_SUMMARY_PROCESSED',
      summary: summarizeProfile(getState().profile),
    });
  };
}

export function profileProcessed(profile) {
  return {
    type: 'PROFILE_PROCESSED',
    profile: profile,
  };
}
