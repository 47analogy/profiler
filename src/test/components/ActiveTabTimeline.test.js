/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import * as React from 'react';
import Timeline from '../../components/timeline';
import { render } from 'react-testing-library';
import { Provider } from 'react-redux';
import { storeWithProfile } from '../fixtures/stores';
import { processProfile } from '../../profile-logic/process-profile';
import { createGeckoProfile } from '../fixtures/profiles/gecko-profile';
import { changeViewAndRecomputeProfileData } from '../../actions/receive-profile';

describe('ActiveTabTimeline', function() {
  const browsingContextID = 123123;

  it('should be rendered properly from the Timeline component', () => {
    const profile = processProfile(createGeckoProfile());
    const store = storeWithProfile(profile);
    store.dispatch(changeViewAndRecomputeProfileData(browsingContextID));

    const { container } = render(
      <Provider store={store}>
        <Timeline />
      </Provider>
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
