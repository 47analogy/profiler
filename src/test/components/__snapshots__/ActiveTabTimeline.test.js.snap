// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActiveTabTimeline should be rendered properly from the Timeline component 1`] = `
<div
  class="timelineSelection"
>
  <div
    class="timelineRuler"
  >
    <ol
      class="timelineRulerContainer"
    />
  </div>
  <div
    class="overflowEdgeIndicator timelineOverflowEdgeIndicator"
  >
    <div
      class="overflowEdgeIndicatorEdge topEdge"
    />
    <div
      class="overflowEdgeIndicatorEdge rightEdge"
    />
    <div
      class="overflowEdgeIndicatorEdge bottomEdge"
    />
    <div
      class="overflowEdgeIndicatorEdge leftEdge"
    />
    <div
      class="overflowEdgeIndicatorScrollbox timelineOverflowEdgeIndicatorScrollbox"
    >
      <div
        class="overflowEdgeIndicatorContentsWrapper"
      >
        <ol
          class="timelineThreadList"
        />
      </div>
    </div>
  </div>
  <div
    class="timelineSelectionHoverLine"
    style="visibility: hidden; left: 0px;"
  />
</div>
`;
